File: CarLine.DataCleanUp/Services/Cleanup/RecordExtractor.cs
using MongoDB.Bson;

namespace CarLine.DataCleanUp.Services.Cleanup;

internal static class RecordExtractor
{
    public static Dictionary<string, string> ExtractAllFields(BsonDocument doc)
    {
        var record = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        foreach (var fieldName in doc.Names)
        {
            // Mongo internal id isn't useful for our output and breaks stable-key logic.
            if (fieldName.Equals("_id", StringComparison.OrdinalIgnoreCase))
                continue;

            if (doc.TryGetValue(fieldName, out var val) && !val.IsBsonNull)
                record[fieldName] = BsonValueConverters.ToTrimmedString(val);
            else
                record[fieldName] = string.Empty;
        }

        return record;
    }

    public static Dictionary<string, string> ExtractCsvFields(BsonDocument doc, IReadOnlyList<string> csvHeader)
    {
        var record = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        foreach (var col in csvHeader)
            if (doc.TryGetValue(col, out var val) && !val.IsBsonNull)
                record[col] = BsonValueConverters.ToTrimmedString(val);
            else
                record[col] = string.Empty;

        return record;
    }
}

File: CarLine.DataCleanUp.FSharp/Domain/RecordExtractor.fs
module CarLine.DataCleanUp.FSharp.RecordExtractor

open System
open System.Collections.Generic
open MongoDB.Bson
open CarLine.DataCleanUp.FSharp.BsonValueConverters

let extractAllFields (doc: BsonDocument) : Dictionary<string, string> =
    let record = Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
    for name in doc.Names do
        if not (name.Equals("_id", StringComparison.OrdinalIgnoreCase)) then
            match doc.TryGetValue(name) with
            | true, v when not v.IsBsonNull -> record[name] <- BsonValueConverters.toTrimmedString v
            | _                             -> record[name] <- ""
    record

let extractCsvFields (doc: BsonDocument) (header: ResizeArray<string>) : Dictionary<string, string> =
    let record = Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
    for col in header do
        match doc.TryGetValue(col) with
        | true, v when not v.IsBsonNull -> record[col] <- BsonValueConverters.toTrimmedString v
        | _                             -> record[col] <- ""
    record


